### PR TITLE
HPCC-14123 PULL in a roxie query gives internal error

### PR DESCRIPTION
**Pull request by richardkchapman**
This was a regression in 5.4.0 builds up to and including rc6

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith Please review.

Note that code is a trimmed copy of the code for catch,skip which had similar requirements. I have not tried to common them up.

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -19164,6 +19164,93 @@ IRoxieServerActivityFactory *createRoxieServerCatchActivityFactory(unsigned _id,
...
+        index = 0;
+        CRoxieServerActivity::start(parentExtractSize, parentExtract, paused);
+    }
+
+    virtual void reset()
+    {
+        while (buff.isItem(index))
+            ReleaseRoxieRow(buff.item(index++));
+        buff.kill();
+        started = false;
```
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: trivial: if you reset started (also set in start()), shouldn't you also reset index = 0 to be consistent?  Or reset neither here.

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -19164,6 +19164,93 @@ IRoxieServerActivityFactory *createRoxieServerCatchActivityFactory(unsigned _id,
...
+
+    virtual const void *nextInGroup()
+    {
+        ActivityTimer t(totalCycles, timeActivities);
+        if (!started)
+            pullInput();
+        if (buff.isItem(index))
+        {
+            const void * next = buff.item(index++);
+            if(next)
```
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: trivial: missing space after 'if'

File: roxie/ccd/ccdserver.cpp [hunk 3]
```diff
@@ -19164,6 +19164,93 @@ IRoxieServerActivityFactory *createRoxieServerCatchActivityFactory(unsigned _id,
...
+    }
+
+protected:
+    void pullInput()
+    {
+        bool EOGseen = false;
+        loop
+        {
+            const void * next = input->nextInGroup();
+            buff.append(next);
```
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: trivial: could avoid 2nd NULL being added at end-of-stream, by moving this line to below if/else clause.

**Comment by jakesmith** [comment]
@richardkchapman - minor comments.

File: roxie/ccd/ccdserver.cpp (hunk 3 above)
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: Not sure why I would want to avoid that second NULL...

Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: no compelling reason, other than it is superfluous

Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: It's a more accurate reflection of what is actually going to be streamed out to caller...

**Comment by richardkchapman** [comment]
Repushed to address minor comments

**Comment by jakesmith** [comment]
@richardkchapman - looks good to merge.

**Pull request merged**